src/services/tfjs-service.ts

```diff
@@ -11,14 +11,292 @@ import {Reference, ResourceManager as rmgr} from './resource-manager.js';
+
+  log('Dimensions expanded.');
+  return rmgr.ref(expanded);
+};
+
+/**
+ * Use bilinear-interpolation to resize a batch of images.
+ *
+ * @param {Reference} images A batch of images to resize of rank 4 or 3. If rank 3, a batch of 1 is assumed.
+ * @param {[number, number]} size The new shape `[newHeight, newWidth]` to resize the image to.
+ * @param {boolean} alignCorners  Defaults to False. If true, rescale input by (new_height - 1) / (height - 1),
+ *  which exactly aligns the 4 corners of images and resized images. If false, rescale by new_height / height.
+ *  Treat similarly the width dimension. Optional
+ * @return {Reference} Images with a new shape.
+ */
+const resizeBilinear = async ({images, size, alignCorners}): Promise<Reference> => {
+  const tf = await requireTf();
+
+  log('Bilinear resizing image(s)...');
+  const images_ = rmgr.deref(images);
+  const resized = await tf.image.resizeBilinear(images_, size, alignCorners);
+  log('Resized image(s).');
+
+  return rmgr.ref(resized);
+};
+
+/**
+ * Return a ranked list (of size K) of classes with their associated probabilities.
+ *
+ * @param {Reference} input Prediction tensor. Must specify either `input`, `y`, or `yHat`.
+ * @param {Reference} y Prediction tensor. Must specify either `input`, `y`, or `yHat`.
+ * @param {Reference} yHat Prediction tensor. Must specify either `input`, `y`, or `yHat`.
+ * @param {string[]} labels A mapping between label index and value.
+ * @param {number} topK The total number of labels to return
+ * @return {ClassificationPrediction[]} A list of predictions, complete with `className` and `probability`.
+ */
+const getTopKClasses = async ({input, y, yHat, labels, topK=3}): Promise<ClassificationPrediction[]> => {
+  log('Getting top K classes...');
+  const input_ = rmgr.deref(input || y || yHat) as tf.Tensor2D;
+
+  const softmax = input_.softmax();
+  const values = await softmax.data();
+  softmax.dispose();
+
+  const valuesAndIndices = [];
+  for (let i = 0; i < values.length; i++) {
+    valuesAndIndices.push({value: values[i], index: i});
+  }
+  valuesAndIndices.sort((a, b) => {
+    return b.value - a.value;
+  });
+  const topkValues = new Float32Array(topK);
+  const topkIndices = new Int32Array(topK);
+  for (let i = 0; i < topK; i++) {
+    topkValues[i] = valuesAndIndices[i].value;
+    topkIndices[i] = valuesAndIndices[i].index;
+  }
+
+  const topClassesAndProbs = [];
+  for (let i = 0; i < topkIndices.length; i++) {
+    topClassesAndProbs.push({
+      className: labels[topkIndices[i]],
+      probability: topkValues[i]
+    });
+  }
+  log('found top K classes.');
+  return topClassesAndProbs;
+};
+
+
+Services.register('graph-model', new GraphModel());
+Services.register('layer-model', new LayersModel());
+
+Services.register('preprocess', {
+  normalize,
+  reshape,
+  expandDims,
+});
+
+Services.register('postprocess', {
+  getTopKClasses,
+});
+
+Services.register('tf-image', {
+  resizeBilinear,
+  imageToTensor,
 });
 
+
```

Comment:
It's strange to have a single import register multiple services. The registration mechanism allows services to be pluggable: if some set of behaviors is atomic, that's one service.

Reply:
I'm a bit confused. Does this imply that this file should be broken up into smaller files, where there's only one register per file?

Reply:
I think he's in a way asking, are all of these really separate services, or are they essentially all one "TF" service, and these are various calls/interactions one might have with it.